tinytodo/src/context.rs

```diff
@@ -355,27 +363,126 @@ impl AppContext {
 
     #[tracing::instrument(skip(policy_set))]
     fn update_policy_set(&mut self, policy_set: PolicySet) -> Result<AppResponse> {
-        let policies = rename_from_id_annotation(policy_set)?;
-        self.policies = policies;
-        info!("Reloaded policy set");
+        let mut new_policies = rename_from_id_annotation(policy_set)?;
+        let mut err = false;
+        // for each existing template-linked policy,
+        //   link against the new version of the template in the new policy set if present
+        for p in self.policies.policies() {
+            match p.template_id() {
+                None => (), // not a template-linked policy
+                Some(tid) => {
+                    // template-linked policy
+                    match new_policies.template(tid) {
+                        None => {
+                            // template not in new policy set
+                            let tidx = tid.clone();
+                            let pidx = p.id().clone();
+                            err = true;
+                            error!("Error when reloading policies: Could not find policy template {tidx} to link {pidx}")
+                        }
+                        Some(_) => {
+                            // found template in new policy set
+                            match p.template_links() {
+                                None => error!("Error when reloading policies: Template with no matching links"),
+                                Some(vals) => {
+                                    // link against new template, using the same policy ID as the old one
+                                    new_policies.link(tid.clone(), p.id().clone(), vals)?
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        // no error during relinking; now validate policies
+        if !err {
+            let validator = Validator::new(self.schema.clone());
+            let output = validator.validate(&new_policies, ValidationMode::default());
+            if !output.validation_passed() {
+                for e in output.validation_errors() {
+                    error!("Error validating linked policies: {e}")
+                }
+            } else {
+                self.policies = new_policies;
+                info!("Reloaded policy set")
+            }
+        }
```

Comment:
I think the `err == true` and `!validation_passed()` cases need to return a `Err` result to abort the update.

Reply:
I don't think so. The actual update only happens when line 406, `self.policies = new_policies;` is run; if it's not run then the current policy set remains intact, regardless of what the handler returns. I've confirmed this with (interactive) testing.